include/pybind11/pybind11.h

```diff
@@ -2612,13 +2612,28 @@ void print(Args &&...args) {
 }
 
 error_already_set::~error_already_set() {
-    if (m_type) {
-        gil_scoped_acquire gil;
-        error_scope scope;
-        m_type.release().dec_ref();
-        m_value.release().dec_ref();
-        m_trace.release().dec_ref();
+    if (!(m_type || m_value || m_trace)) {
+        return; // Avoid gil and scope overhead if there is nothing to release.
+    }
+#if PY_VERSION_HEX >= 0x03070000
+    if (PyGILState_Check() == 0 && _Py_IsFinalizing() != 0) {
```

Comment:
This check is still racy.  For example, we can have the sequence:

1. Thread 2 is here and finds `_Py_IsFinalizing() != 0`
2. Main thread starts the late stage of finalization and causes `_Py_IsFinalizing()` to become non-zero.
3. Thread 2 uses `gil_scoped_acquire_simple` and crashes.

I don't believe there is any safe way to use Py_IsFinalizing as you are trying to use it here --- the problem is that you need a way to atomically check that `Py_IsFinalizing` is false *and* acquire the GIL, and furthermore ensure that `_Py_IsFinalizing` remains false until the end of your block.  The Python C API doesn't provide any way to do that directly at the moment --- I addressed that in https://github.com/python/cpython/pull/28525 but unfortunately I haven't heard back from the Python maintainers regarding that PR in a long time.

It is possible to implement a workaround, as is done in TensorStore:

https://github.com/google/tensorstore/blob/fe7aae44a788dbdfc731ea1c00b83f0d51e4f2f4/python/tensorstore/gil_safe.h#L80

It is based on registering an atexit handler so that we can delay finalization while in such a block.

Reply:
Note: The reason that we need to prevent finalization from starting while we are still in the block is that even after acquiring the GIL, if we call Python APIs, those Python APIs may directly or indirectly release and then re-acquire the GIL.  If finalization starts after the GIL is released but before it is re-acquired, then we have the same problem.